api/client.go

```diff
@@ -172,6 +173,15 @@ func (c *Client) newRequest(
 
 	req.Header.Add("User-Agent", c.conf.UserAgent)
 
+	// If our context has a timeout/deadline, tell the server how long is remaining.
+	// This may allow the server to configure its own timeouts accordingly.
+	if deadline, ok := ctx.Deadline(); ok {
+		ms := deadline.Sub(time.Now()).Milliseconds()
+		if ms > 0 {
+			req.Header.Add("Buildkite-Timeout-Milliseconds", strconv.FormatInt(ms, 10))
```

Comment:
Is `Buildkite-Timeout-Milliseconds` a good header name?

Is putting the units in the header name good? It makes parsing the value much simpler.

Is milliseconds a good time granularity? It was either that or seconds, which would probably be adequate but maybe not? And opened more (minor) questions about integer-rounding etc.

I also considered side-stepping the precision question and making the name shorted with `Buildkite-Timeout` and ISO 8601 duration value e.g. `PT10S` (10 seconds) but I couldn't see a good stdlib formatter for those, and it makes parsing it less simple, and I don't think they're _very_ widely used nor very pretty/readable 😬

I've specifically avoided the [deprecated](https://datatracker.ietf.org/doc/html/rfc6648) `X-` prefix, which we use in one or two other places. I'd rather (gradually, one day) drop the `X-` from those vs adding it here for consistency. (/cc @triarius 😁)

Reply:
If I were in control of the other end of the request, I'd be chucking the value through something like `time.ParseDuration`, which can interpret suffixes like `ms`. So my vote would be for `Buildkite-Timeout: 123ms`. `Buildkite-Timeout-Milliseconds` is otherwise fine with me.

Reply:
Makes sense to not have `X-`, I very much regret introducing it 😅

There seems to be an assumption of seconds for various headers:

`Retry-After` is either a [date or in seconds](https://datatracker.ietf.org/doc/html/rfc7231#section-7.1.3). `Cache-Control` seems to assume seconds for its sub-values as well. `Age` [assumes seconds](https://datatracker.ietf.org/doc/html/rfc9111#field.age).

That's very confusing, though. I prefer the units in the header name approach to that. It is possible to put the units in the header value, i.e.
```
Buildkite-Timeout: 100ms
```
And we could default to seconds if there is no unit.

But my feelings are not that strong on this matter.

Reply:
^ I guess @DrJosh9000 and I have the same position

Reply:
Yeah… I'd like to use a common encoding of the duration/units, but…

```ruby
ActiveSupport::Duration.parse("100ms")
# ActiveSupport::Duration::ISO8601Parser::ParsingError: Invalid ISO 8601 duration: "100ms"
```

Ruby / Rails / ActiveSupport likes ISO 8610, Go stdlib likes a nicer but non-standard form, so I reckon `Buildkite-Timeout-Milliseconds` with an integer value ends up simplest 😢

Reply:
Hah:
<img width="719" alt="image" src="https://github.com/buildkite/agent/assets/15759/b19b7b66-8150-43f6-96aa-3ffbe75f1c96">

— https://stackoverflow.com/questions/657309/how-to-parse-days-hours-minutes-seconds-in-ruby

Hello twelve-years-ago-@sj26 👋🏼

Reply:
(chronic_duration hasn't been updated for about a decade, GitLab have a fork https://rubygems.org/gems/gitlab_chronic_duration which hasn't been updated for 4 years, and I'd be inclined to not add libraries just for this anyway)